Speed up the __hash__ method of DiscreteFactor

reeingal opened the pull request:
The `__hash__()` method of `DiscreteFactor` use `str(phi.values)` to
construct the hash value.

The `phi.values` is a `numpy.ndarray` object, and `str(phi.values)`
will spend too much time when the shape of `numpy.ndarray` is huge.

I find that `hash(phi.values.data.tobytes())` is much faster,
so I suggest to use `str(hash(phi.values.data.tobytes()))` to construct
the hash value, which will speed up the execute time like `predict()` of
`BayesianNetwork` for large networks.

### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [ ] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #1525

### List of changes to the codebase in this pull request
- Modify the `__hash__()` method of `DiscreteFactor` to speed up execute time.

reeingal committed:
Speed up the __hash__ method of DiscreteFactor

The `__hash__()` method of `DiscreteFactor` use `str(phi.values)` to
construct the hash value.

The `phi.values` is a `numpy.ndarray` object, and `str(phi.values)`
will spend too much time when the shape of `numpy.ndarray` is huge.

I find that `hash(phi.values.data.tobytes())` is much faster,
so I suggest to use `str(hash(phi.values.data.tobytes()))` to construct
the hash value, which will speed up the execute time like `predict()` of
`BayesianNetwork` for large networks.

ankurankan commented on pgmpy/factors/discrete/DiscreteFactor.py:
```diff
@@ -1026,7 +1026,7 @@ def __hash__(self):
         return hash(
             str(sorted_var_hashes)
-            + str(phi.values)
+            + str(hash(phi.values.data.tobytes()))
```
@reeingal `phi.values.tobytes()` seems to be slightly faster than `phi.values.data.tobytes()`.
```python
In [20]: %timeit data.data.tobytes()
22.4 µs ± 38.7 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [21]: %timeit data.tobytes()
21.4 µs ± 32.9 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [22]: data.tobytes() == data.data.tobytes()
Out[22]: True
```

ankurankan commented on pgmpy/factors/discrete/DiscreteFactor.py:
```diff
@@ -1026,7 +1026,7 @@ def __hash__(self):
             str(sorted_var_hashes)
-            + str(phi.values)
+            + str(hash(phi.values.data.tobytes()))
             + str(phi.cardinality)
```
Can also convert `phi.cardinality` to bytes.

reeingal replied:
I am a new Github user, although signed up early. I modify the code according to your suggestion, and commit again, is this right?

ankurankan replied:
Yes, this is good.

reeingal committed:
Use numpy.ndarray.tobytes() to generate hash value

Use `tobytes()` method  of `numpy.ndarray` object will get the same
result as `numpy.ndarray.data.tobytes()` and run more faster.

Change `str(phi.cardinality)` to `str(hash(phi.cardinality.tobytes()))`
in the hash caculation of  `DiscreteFactor`.
